### DISCUSSION TITLE
NutUI-React  共建指南

### DISCUSSION THREAD

**Discussion by oasis-cloud** [jdf2e/nutui-react#348] [Announcements]
# NutUI-React 组件共建指南

我们非常欢迎社区的开发者向 NutUI 做出贡献。在提交贡献之前，请花一些时间阅读以下内容，保证贡献是符合规范并且能帮助到社区。

## @nutui/nutui-react 的开发配置

你需要保证你的 Node.js 版本大于 16，启动 taro 版本的 demo，需要 taro 3.6.x。把仓库 clone 到本地，并运行以下命令：

```bash
pnpm install
pnpm dev
```

## @nutui/nutui-react-taro 的开发配置

```bash
pnpm dev:taro:weapp
pnpm run dev:taro:h5
```

## 提交 commit
整个 NutUI 仓库遵从 Angular Style Commit Message Conventions，在输入 commit message 的时候请务必遵从此规范。

Title Format type(ComponentName?)：commit message

例如：

docs: fix type in quickstart build: optimize build speed fix(Button): incorrect style feat(Button): add color prop

允许的类型 Types:

upd chore docs feat fix test refactor revert style releas

## Pull Request 指南
- 务必保证 `pnpm build` `pnpm build:taro` 能够编译成功；
- 当相关包的 package.json 含有 pnpm test 命令时，必须保证所有测试用例都需要通过；
- 当相关包有测试用例时，请给你提交的代码也添加相应的测试用例；
- 提交代码 commit 时，commit 信息需要遵循 Angular Style Commit Message Conventions。
- 如果提交的代码非常多或功能复杂，可以把 PR 分成几个 commit 一起提交。我们在合并时会根据情况 squash。

## Credits
感谢以下所有给 NutUI 贡献过代码的开发者

https://github.com/jdf2e/nutui-react/graphs/contributors

**Comment by 5201314999** [comment]
务必保证 npm run build npm run build:taro 能够编译成功；npm run build 类型的命令需要使用 node v17 这个不理解

**Comment by xiaoyatong** [comment]
# NutUI-React 常见问题 (FAQ)

## Q: 为什么我在 `config.json` 中配置了新组件，或者拉取了含有新组件的分支后，本地运行 `dev` 或 `dev:taro h5` 时，Demo 里无法展示该组件（或报 `undefined` 错误）？

> [!WARNING]
> 当遇到如下情况时，通常是本问题引起的：
> - 侧边栏菜单点击新组件，右侧手机模拟器白屏或展示为空。
> - 控制台出现类似 `Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined.` 的 React 报错。
> - 在 `demo1.tsx` 中导入的组件打印出来为 `undefined`。

### 原因分析

NutUI-React 的项目架构中，负责导出组件的总入口文件以及生成文档路由的文件是**由脚本动态生成的**，并非手动维护。为了减少合并冲突，这些文件被添加到了 `.gitignore` 列表中，不会被 Git 追踪。

主要包含以下几个核心的生成文件：
* **H5 端组件入口**: `src/packages/nutui.react.ts`
* **Taro 端组件入口**: `src/packages/nutui.react.taro.ts`
* **文档与路由配置**: `src/sites/doc/docs.ts` 及 `docs.taro.ts` 等

当你新增了一个组件（并在 `config.json` 中声明），或者 `git pull` 拉取了团队其他成员提交的新组件代码时，**你本地电脑上的这些动态文件依然是旧的**。由于旧的文件里没有写入新组件的 `export` 语句，导致你在 `demo.tsx` 等地方通过 `@nutui/nutui-react` 引入该组件时，拿到的是 `undefined`。

### 解决办法

你需要在本地手动触发一次脚本，重新根据最新的 `config.json` 生成这些入口和路由文件。

> [!TIP]
> **推荐的终极解决办法（一键修复所有环境）：**
> 在项目根目录下执行以下命令，它会同时重新生成 H5 和 Taro 的所有入口与页面映射文件：
> ```bash
> pnpm run prepare
> ```

如果你只想针对当前正在开发的特定环境进行生成，也可以拆分执行：

**仅修复 H5 环境：**
```bash
pnpm run generate:file
# 等同于执行 node scripts/generate-nutui.js
```

**仅修复 Taro 环境：**
```bash
pnpm run generate:file:taro
pnpm run generate:file:taro:pages
# 等同于执行 node scripts/taro/generate-nutui-taro.js 和 generate-taro-pages.js
```

**总结规则**：**只要修改了 `config.json` 或拉取了含有组件结构变动的代码，请务必执行一次 `pnpm run prepare` 以刷新本地生成的入口文件。**
